Point_set_processing_3/include/CGAL/structure_point_set.h

```diff
@@ -0,0 +1,1483 @@
+    }
+
+    std::cerr << "Computing corner directions... " << std::endl;
+#endif
+      
+    compute_corner_directions (epsilon);
+
+#ifdef CGAL_PSP3_VERBOSE
+    std::cerr << " -> Done" << std::endl;
+    std::cerr << "Refining sampling... " << std::endl;
+#endif
+      
+    refine_sampling (epsilon);
+
+#ifdef CGAL_PSP3_VERBOSE
+    std::cerr << " -> Done" << std::endl;
+
+    std::cerr << "Cleaning data set... " << std::endl;
+#endif
+      
+    clean ();
+
+#ifdef CGAL_PSP3_VERBOSE
+    std::cerr << " -> Done" << std::endl;
+#endif
+  }
+
+  void project_inliers ()
+  {
+    for(std::size_t i = 0; i < m_planes.size (); ++ i)
+      for (std::size_t j = 0; j < m_planes[i]->indices_of_assigned_points ().size(); ++ j)
+        {
+          std::size_t ind = m_planes[i]->indices_of_assigned_points ()[j];
+          m_points[ind] = static_cast<Plane> (*(m_planes[i])).projection (m_points[ind]);
+        }
+  }
+
+  void resample_planes (double epsilon)
+  {
+    double grid_length = epsilon * (std::sqrt(2.) - 1e-3);
+
+    for (std::size_t c = 0; c < m_planes.size (); ++ c)
+      {
+        //plane attributes and 2D projection vectors
+        Plane plane = static_cast<Plane> (*(m_planes[c]));
+        Vector vortho = plane.orthogonal_vector();
+        Vector b1 = plane.base1();
+        Vector b2 = plane.base2();
+			
+        b1 = b1 / std::sqrt (b1 * b1);
+        b2 = b2 / std::sqrt (b2 * b2);
+
+        std::vector<Point_2> points_2d;
+
+        //storage of the 2D points in "pt_2d"
+        for (std::size_t j = 0; j < m_planes[c]->indices_of_assigned_points ().size(); ++ j)
+          {
+            std::size_t ind = m_planes[c]->indices_of_assigned_points ()[j];
+            const Point& pt = m_points[ind];
+            points_2d.push_back (Point_2 (b1.x() * pt.x() + b1.y() * pt.y() + b1.z() * pt.z(),
+                                          b2.x() * pt.x() + b2.y() * pt.y() + b2.z() * pt.z()));
+          }
+
+
+        //creation of a 2D-grid with cell width = grid_length, and image structures
+        CGAL::Bbox_2 box_2d = CGAL::bbox_2 (points_2d.begin(), points_2d.end());
+        std::size_t Nx = (std::size_t)((box_2d.xmax() - box_2d.xmin()) / grid_length) + 1;
+        std::size_t Ny = (std::size_t)((box_2d.ymax() - box_2d.ymin()) / grid_length) + 1;
+          
+        std::vector<std::vector<bool> > Mask (Nx, std::vector<bool> (Ny, false));
+        std::vector<std::vector<bool> > Mask_border (Nx, std::vector<bool> (Ny, false));
+        std::vector<std::vector<std::vector<std::size_t> > >
+          point_map (Nx, std::vector<std::vector<std::size_t> > (Ny, std::vector<std::size_t>()));
+
+        //storage of the points in the 2D-grid "point_map"
+        for (std::size_t i = 0; i < points_2d.size(); ++ i)
+          {
+            std::size_t ind_x = (std::size_t)((points_2d[i].x() - box_2d.xmin()) / grid_length);
+            std::size_t ind_y = (std::size_t)((points_2d[i].y() - box_2d.ymin()) / grid_length);
```

Comment:
I think @lrineau mentioned that we should avoid using C-style cast to avoid potential mistake that are detected with a flag of gcc I forgot.